### DISCUSSION TITLE
Encryption key in plain text in backup directory.  Security issue?

### DISCUSSION THREAD

**Discussion by GalacticSun** [borgbase/vorta#2092] [FAQ]
I'm new to Vorta and just trying to learn more about the encryption keys.  When setting up my profile I opted to use a passphrase and an encryption key ("repokey/blake2") and it created an entry in the config file in the backup directory that contains that key, in plain text.  It appears that Vorta needs that key in that file in that location to operate.  If the backup dir is on a portable USB device and that device is stolen, then they have the encryption key.  Isn't that a security concern?  Thanks.

**Comment by goebbe** [comment]
@GalacticSun My understanding is that the repokey itself is virtually useless without the passphrase, i.e. you cannot decrypt the backup with the repokey only. If you use a passphrase, you always need both, the **repokey _and_ the passphrase**. 

Since Vorta/ borg use strong encryption, it is considered save, to use repokey (i.e. storing the keyfiles with the repository) in combination with a strong passphrase. This is why Vorta uses repokey by default. 

Obviously, you have to **use a strong passphrase**!  

Note that you could also opt for keeping the encryption key in your home directory: Choose "Keyfile-Blake2 (Key stored in home directory)", in the advanced options, when creating a new repository.  But then you should take care to not delete/ lose your keyfile, since it would be impossible to decrypt your backup without it.

**Reply by GalacticSun** [reply]
Thanks, @goebbe, then maybe I should use the option to have Vorta put the key in my home directory and then back it up to another location.  That sounds like a much more secure option to me.

If an attacker were to get access to the backup directory (on USB, etc) and get the plain text key, then that is 50% of the security circumvented right there, so I still think having the key in plain text in the backup directory doesn't make any sense at all, especially as the default option.

It's like taping your house key to your front door but then asking for a passcode on another lock on the door.  It kinda renders the key useless since it's right there.

Shouldn't there be better documentation explaining this?  I couldn't find anything addressing this specific topic and how to learn the details.  In fact, I only found the instructions for exporting the encryption key in the log file after I ran my first backup - which I likely would have missed if I hadn't been curious as to what the log said.  But exporting the key is pointless anyway, because Vorta requires the key to be in the backup directory to even run!

Knowing how to securely store and handle encryption keys is vital for new users.  Hopefully this can be made more clear in the UI and the docs.

**Reply by goebbe** [reply]
With a strong passphrase, the repokey itself is useless. This is why Vorta can store the key with the repository.  :-) 

Figuratively speaking: Think about a vault with a very, very long double sided security key, where one side is represented by the repokey and the other side is the passphrase. Knowing only one side of this key will not open the vault. 

Would it be even more secure if you would not even know the repokey - yes it would - but since knowing only one side of the key will not open the vault anyway, the difference in actual security will not be as big as you might think. 

Either way, it is your choice. If you don't think that passphrase security is enough, go with a local keyfile. :-)

**Reply by GalacticSun** [reply]
Well, but that's not the point of an encryption key.  An encryption key is supposed to be another (and major) factor in the security of the system it is used in.  By saying the key is useless only proves my point: that storing the key in the repo is of no security benefit.  Therefore, why do it?  Why don't the devs address this?

Personally, I have switched to a local keyfile (in /home, and backed up elsewhere), but this begs the question why are the devs doing this and allowing (and recommending!) the key be stored in the repo.  At best it's a false sense of security, and at worst it is a security flaw.  There should at least be some documentation explaining this security issue and an effort to change it.  Don't you think?  Am I missing something here?

**Reply by goebbe** [reply]
I believe it all comes down to the question, "How secure is secure enough?" Is "extremely secure" good enough? Or should it be "paranoic secure"? On the same line of argument, we could ask if we should force users to use very long and complicated (random) passphrases. This would not be convenient, but very secure. 
By the way, a third option would be to only use a local encryption key, without a passphrase. Even this option would likely be more secure than the average passphrase. :-) 

If you believe that the defaults are not sensible, feel free to open an issue. I think you have a point and it's worth a discussion.

**Reply by GalacticSun** [reply]
Thanks, I appreciate our discussion and your points.  I agree that it's up to the user and "how secure is secure".  True enough.  I guess my concern is more for newbies to Vorta/Borg, like myself, and understanding its security.  In providing a key like they do, though, you'd think it would be an actually useful security measure and not just window dressing.

I also agree that forcing people into "better security" is not right.  But educating them is, for sure.  Do you think this is a Vorta or a Borg thing?

**Reply by goebbe** [reply]
Passphrase protection is probably well understood by a large share of Vorta users. So presumably everybody understands that it is crucial to use a secure passphrase. If you want it even more secure, Vorta provides the option of using local keyfiles - but this is considered as advanced use case - where the user should know what he is doing: 
- it is impossible to memorize a keyfile, but you could easily build a long passphrase that can be memorized
  - therefore, if you lose/ delete/ destroy your keyfile, there is no way to access/ restore the backups
- the keyfile is probably saved somewhere in your home directory, everybody, who has access to your home directory can read/ copy the keyfile. If your system is not correctly configured, other users on your computer could read your keyfile too.

My feeling is that Vorta / borg users are expected to read the docs, at least for advanced use cases.

By the way, everybody is free to write his private blog about how to harden Vorta or how to use local keyfiles etc.  ;-)   And in a way, by discussion this topic openly, we are just about contributing to the (informal) documentation.

**Reply by GalacticSun** [reply]
I agree that passphrase protection is well understood, but I don't think encryption keys are as well understood.

Encryption keys do provide enhanced security, and I think it is up to the devs to provide proper guidance in their docs as to how they implemented them.  But I don't think they have, unfortunately.  And, by putting 'repokey' as the default option that it's a less secure choice as it exposes the key in the backup repository, which, as I've pointed out, appears to me to be a security flaw, compromising 50% of one's security if they are using a key, essentially rendering a key useless, thus invalidating the idea that encryption keys provide added security.

Yes, Vorta users are expected to read the docs, which brings me back to an earlier point I made: regarding this issue, I haven't seen any.

And, sure, I could blog about it, but I don't have or want a tech blog.  This is entirely on the devs, imho, since it's their product (albeit free).  Hopefully this discussion can raise some awareness about this issue, unless I'm missing something about encryption keys (and I very well may be!).  ;)

**Reply by goebbe** [reply]
@GalacticSun May I ask questions about your setup? When creating a repo with _local_ keyfiles, they are saved in ~/.config/borg/keys by default, right? Do you keep a backup of the keyfile, somewhere? 
How would you proceed, if you want to restore on a different computer? Would it be sufficient to copy/ past the keyfile into ~/.config/borg/keys on that computer - or would you use "borg import key"?

**Reply by GalacticSun** [reply]
Hi @goebbe.  It depends on what you're using (Vorta, Borg, or other Borg front-end) and how you installed it (Flatpak, apt, etc).  But, yes, you'd create a backup of your keyfile somewhere.  Then use it to restore when you want to.  But if you're just mounting the repo to copy some files over it can use your local keyfile and is pretty seamless.
As for restoring on another computer, I think just copying and pasting it is sufficient.  Haven't used "borg import key".  But I'm new to Borg, too. ;)

**Comment by goebbe** [comment]
Here is a related question and the answers from borg devs:  https://github.com/borgbackup/borg/issues/5285

**Reply by GalacticSun** [reply]
Thanks for finding that, @goebbe, that's good info!

I think [this answer](https://github.com/borgbackup/borg/issues/5285#issuecomment-670021910) should be (albeit cleaned up a bit) in [the docs](https://vorta.borgbase.com/usage/).  A section in the docs covering keyfiles would be very helpful for those interested in using repokeys/keyfiles.  It's not beneficial to leave out docs that cover the "Advanced" tab because using the options in the "Advanced" tab might be considered to be "Advanced".  Even a version of the correlating explanation from the post I linked to above would be sufficient.

And the post I linked to could be expanded to explain the pros and cons of each scenario with some examples.  Such as: "If you use an external USB HD for backups, don't use the repokey because if your external HD gets stolen, 50% of your security is compromised because the attacker has the key", etc.  I just think things could and should be made a _whole_ lot clearer in the docs.  Especially for newbies.

**Reply by goebbe** [reply]
For general considerations about the security of passphrases / keyfile one could also **refer to outside sources** (if you know good sources that explain why we should use keyfiles instead of passphrases). Most of this would likely apply to many different scenarios. 
If you want to push **changes for the docs**, you could open an issue, to ask the devs, if they would support adding some changes to the docs.  The more specific your request, the higher the chances that somebody will look into this. 
If the devs agree, somebody has to **propose the changes to the docs** (pull request). This is free software - feel free to make a proposal, if nobody else cares / has the time to work on the docs.

**Reply by GalacticSun** [reply]
Yup, I understand.  Anyone can look into outside sources, but if you're going to develop free software and provide docs, I'd think they'd want to provide a bit more comprehensive docs.  In this case it is simple to do.  It constitutes just a few paragraphs, and it's basically already been written.

Yeah, I suppose if I want to suggest the change, I will.  It just took me far too long to come to an understanding and a conclusion about this when all they had to do was put a paragraph in the docs summarizing everything.  Hopefully this can save the next person a bit of time and confusion regarding using keys with this (excellent) software! ;)
